extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/datalink/MongoDataLink.java

```diff
@@ -0,0 +1,143 @@
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.jet.mongodb.datalink;
+
+import com.hazelcast.config.DataLinkConfig;
+import com.hazelcast.datalink.DataLink;
+import com.hazelcast.datalink.DataLinkResource;
+import com.hazelcast.datalink.impl.ReferenceCounter;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoDatabase;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.internal.util.Preconditions.checkNotNull;
+import static com.hazelcast.internal.util.Preconditions.checkState;
+
+/**
+ * Creates a MongoDB DataLink.
+ *
+ * According to {@link MongoClient} documentation, the client object is responsible for maintaining connection pool,
+ * so this data link just returns one, cached client.
+ *
+ * @since 5.3
+ */
+public class MongoDataLink implements DataLink {
+    /**
+     * Name of a property which holds connection string to the mongodb instance.
+     */
+    public static final String CONNECTION_STRING_PROPERTY = "connectionString";
+    protected final DataLinkConfig config;
+    protected final ReferenceCounter refCounter;
+    private MongoClient mongoClient;
+
+    /**
+     * Creates a new data link based on given config.
+     */
+    public MongoDataLink(DataLinkConfig config) {
+        this.config = config;
+        this.refCounter = new ReferenceCounter(this::destroy);
+
+        String connectionString = config.getProperty(CONNECTION_STRING_PROPERTY);
+        checkNotNull(connectionString, "connectionString property cannot be null");
+        this.mongoClient = new CloseableMongoClient(MongoClients.create(connectionString), refCounter::release);
+    }
+
+    /**
+     * Returns an instance of {@link MongoClient}.
+     *
+     * {@linkplain MongoClient#close()} won't take an effect if there will be still some usages of given client.
+     */
+    @Nonnull
+    public MongoClient getClient() {
+        refCounter.retain();
+        checkState(mongoClient != null, "Mongo client should not be closed at this point");
+        return mongoClient;
+    }
+
+    /**
+     * Returns name of this data link.
+     */
+    @Nonnull
+    @Override
+    public String getName() {
+        return config.getName();
+    }
+
+    /**
+     * Lists all MongoDB collections in all databases.
+     */
+    @Nonnull
+    @Override
+    public List<DataLinkResource> listResources() {
+        List<DataLinkResource> resources = new ArrayList<>();
+        for (String databaseName : mongoClient.listDatabaseNames()) {
+            MongoDatabase database = mongoClient.getDatabase(databaseName);
+            for (String collectionName : database.listCollectionNames()) {
+                resources.add(new DataLinkResource("collection", collectionName));
```

Comment:
what if I have the same collection in different databases?

Reply:
That's actually a good question. Mongo does not use `schema.database.table` naming schema like most DBs do; I could return here `databaseName.collectionName`, but won't it be confusing for the user? or: what will be less confusing ;)